### PR TITLE
feat(watcher): watch solana burns

### PR DESCRIPTION
**Pull request by TheAncientGoat**
This PR introduces support for watching the Solana host chain for burns, in order to ensure they get submitted in case a user is unable to do so.

### CODE REVIEW

**Review by tok-kkk — APPROVED** [review]

File: watcher/watcher.go [hunk 1]
```diff
@@ -117,6 +123,138 @@ func (fetcher EthBurnLogFetcher) FetchBurnLogs(ctx context.Context, from uint64,
...
+}
+
+func (fetcher SolFetcher) FetchBurnLogs(ctx context.Context, from uint64, to uint64) (chan BurnLogResult, error) {
+	resultChan := make(chan BurnLogResult)
+
+	go func() {
+		func() {
+			defer close(resultChan)
+			for i := from; i < to; i++ {
+				nonce := i + 1
```
Comment (jazg) [inline watcher/watcher.go:154]: Can you explain what this nonce is tracking?

File: watcher/watcher.go [hunk 2]
```diff
@@ -117,6 +123,138 @@ func (fetcher EthBurnLogFetcher) FetchBurnLogs(ctx context.Context, from uint64,
...
+				// Fetch account data at gateway registry's state
+				accountInfo, err := fetcher.client.GetAccountInfo(ctx, programPubk)
+				if err != nil {
+					resultChan <- BurnLogResult{Error: fmt.Errorf("getting burn log data for burn: %v err: %v", i, err)}
+					return
+				}
+				data := accountInfo.Value.Data
+
+				if len(data) != 41 {
+					resultChan <- BurnLogResult{Error: fmt.Errorf("deserializing burn log data: %v", err)}
```
Comment (jazg) [inline watcher/watcher.go:179]:
```suggestion
					resultChan <- BurnLogResult{Error: fmt.Errorf("deserializing burn log data: expected data len 41, got %v", len(data))}
```

File: watcher/watcher.go [hunk 3]
```diff
@@ -117,6 +123,138 @@ func (fetcher EthBurnLogFetcher) FetchBurnLogs(ctx context.Context, from uint64,
...
+		gatewayStatePubk: programPubk,
+		gatewayAddress:   gatewayAddress,
+	}
+}
+
+func (fetcher SolFetcher) FetchBurnLogs(ctx context.Context, from uint64, to uint64) (chan BurnLogResult, error) {
+	resultChan := make(chan BurnLogResult)
+
+	go func() {
+		func() {
```
Comment (jazg) [inline watcher/watcher.go:151]: I think we can get rid of this inner function.

**Review by jazg — CHANGES_REQUESTED** [review]
Looks good! Just left a couple comments.

File: watcher/watcher.go (hunk 1 above)
Comment (TheAncientGoat) [reply watcher/watcher.go]: Added a comment in the code; it refers to the number of burns on the solana gateway (which starts indexing at 1)

Comment (jazg) [reply watcher/watcher.go]: Ah, I assumed `i` was just the block number in this case.

Comment (TheAncientGoat) [reply watcher/watcher.go]: Nope. Solana scans over burn indexes, Ethereum scans block numbers, and then iterates over events (nonce is still technically the burn index)

Comment (jazg) [reply watcher/watcher.go]: Makes sense.

**Review by jazg — APPROVED** [review]

**Pull request merged**